File: tests/Infrastructure/EventPublisherTest.php
<?php

namespace Test\Infrastructure;

use App\Domain\IDomainEvent;
use App\Infrastructure\EventPublisher;
use Illuminate\Events\Dispatcher;

class EventPublisherTest extends \PHPUnit_Framework_TestCase {

    public function testGivenAnEventSubscriber_WhenPublishEvent_ThenSubscriberIsCalled() {
        $eventPublisher = new EventPublisher(new Dispatcher());
        $eventA = new EventA();
        $subscriberCalled = false;
        $eventPublisher->subscribe(get_class($eventA), function() use(&$subscriberCalled) {
            $subscriberCalled = true;
        });

        $eventPublisher->publish($eventA);

        \Assert\that($subscriberCalled)->true();
    }
}

class EventA implements IDomainEvent {

}

File: Mixter.Infrastructure.Tests/EventPublisherTest.cs
using Mixter.Domain;
using NFluent;
using Xunit;

namespace Mixter.Infrastructure.Tests
{
    public class EventPublisherTest
    {
        [Fact]
        public void GivenHandlerWhenPublishThenCallHandler()
        {
            var handler = new EventHandler<EventA>();
            var publisher = new EventPublisher();
            publisher.Subscribe(handler);

            publisher.Publish(new EventA());

            Check.That(handler.IsCalled).IsTrue();
        }

        [Fact]
        public void GivenDifferentHandlersWhenPublishThenCallRightHandler()
        {
            var handlerB = new EventHandler<EventB>();
            var handlerA = new EventHandler<EventA>();
            var publisher = new EventPublisher();
            publisher.Subscribe(handlerA);
            publisher.Subscribe(handlerB);

            publisher.Publish(new EventB());

            Check.That(handlerA.IsCalled).IsFalse();
            Check.That(handlerB.IsCalled).IsTrue();
        }

        private class EventHandler<TEvent> : IEventHandler<TEvent> where TEvent : IDomainEvent
        {
            public bool IsCalled { get; private set; }

            public void Handle(TEvent evt)
            {
                IsCalled = true;
            }
        }

        private class EventA : IDomainEvent
        {
            public object GetAggregateId()
            {
                return "A";
            }
        }

        private class EventB : IDomainEvent
        {
            public object GetAggregateId()
            {
                return "B";
            }
        }
    }
}
